git_wrapper/commit.py

```diff
@@ -0,0 +1,135 @@
+
+import git
+
+from git_wrapper.base import GitWrapperBase
+from git_wrapper import exceptions
+from git_wrapper.utils.decorators import reference_exists
+
+
+logger = logging.getLogger(__name__)
+
+
+class GitWrapperCommit(GitWrapperBase):
+    """Provides git commit functionality"""
+
+    def commit(self, message):
+        """Create a commit for changes to tracked files in the repo.
+           Equivalent to `git commit -a -m <message>`.
+
+           :param str message: The commit message
+        """
+        if not message:
+            raise exceptions.CommitMessageMissingException()
+
+        # Add tracked files to the index
+        self.repo.git.add(update=True)
+        changes = self.repo.git.diff(name_only=True, staged=True)
+        if not changes:
+            logger.info("No changes to commit.")
+            return False
+
+        # Commit the changes
+        logger.debug("Preparing to commit changes to the following files: %s" % changes)
+        commit = self.repo.index.commit(message)
+        logger.debug("Committed changes as commit %s" % commit.hexsha)
+
+    @reference_exists('branch_name')
+    def apply_patch(self, branch_name, path):
+        """Apply a git patch file on top of the specified branch.
+
+           :param str branch_name: The name of the branch or reference to apply the patch to
+           :param str path: Path to a git-formatted patch file (cf. git format-patch)
+        """
+        # Expand file (also needed for git-am) and check it exists
+        full_path = os.path.expanduser(path)
+        if not os.path.isfile(full_path):
+            raise exceptions.FileDoesntExistException('%s is not a file.' % full_path)
+
+        # Checkout
+        try:
+            self.repo.git.checkout(branch_name)
+        except git.GitCommandError as ex:
+            msg = "Could not checkout branch %s. Error: %s" % (branch_name, ex)
+            raise exceptions.CheckoutException(msg) from ex
+
+        # Apply the patch file
+        try:
+            self.repo.git.am(full_path)
+        except git.GitCommandError as ex:
+            msg = "Could not apply patch %s on branch %s. Error: %s" % (path, branch_name, ex)
+            raise exceptions.ChangeNotAppliedException(msg) from ex
+
+    @reference_exists('branch_name')
+    def apply_diff(self, branch_name, diff_path, message):
+        """Apply a diff on top of the specified branch.
+
+           :param str branch_name: The name of the branch or reference to apply the diff to
+           :param str diff_path: Path to the diff file
+           :param str message: Commit message
+        """
+        # Ensure we don't commit more than we mean to
+        if self.repo.is_dirty():
+            msg = "Repository %s contains uncommitted changes. Please clean workspace before proceeding." % self.repo.working_dir
+            raise exceptions.DirtyRepositoryException(msg)
+
+        # Check diff file exists
+        full_path = os.path.expanduser(diff_path)
```

Comment:
Good idea. I left the new function in the Commit wrapper class for now, but it can be extracted higher when it's needed elsewhere. I've been burnt by utils classes created too early before, that ended up becoming horrible catch-all for anything and everything...